internal/factsengine/gatherers/saptune.go

```diff
@@ -0,0 +1,140 @@
+package gatherers
+
+import (
+	"encoding/json"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/trento-project/agent/internal/core/saptune"
+	"github.com/trento-project/agent/pkg/factsengine/entities"
+	"github.com/trento-project/agent/pkg/utils"
+)
+
+const (
+	SaptuneGathererName = "saptune"
+)
+
+// nolint:gochecknoglobals
+var whitelistedArguments = map[string][]string{
+	"status":          {"status", "--non-compliance-check"},
+	"solution-verify": {"solution", "verify"},
+	"solution-list":   {"solution", "list"},
+	"note-verify":     {"note", "verify"},
+	"note-list":       {"note", "list"},
+}
+
+// nolint:gochecknoglobals
+var (
+	SaptuneNotInstalled = entities.FactGatheringError{
+		Type:    "saptune-not-installed",
+		Message: "saptune is not installed",
+	}
+
+	SaptuneVersionUnsupported = entities.FactGatheringError{
+		Type:    "saptune-version-not-supported",
+		Message: "currently installed version of saptune is not supported",
+	}
+
+	SaptuneArgumentUnsupported = entities.FactGatheringError{
+		Type:    "saptune-unsupported-argument",
+		Message: "the requested argument is not currently supported",
+	}
+
+	SaptuneMissingArgument = entities.FactGatheringError{
+		Type:    "saptune-missing-argument",
+		Message: "missing required argument",
+	}
+
+	SaptuneCommandError = entities.FactGatheringError{
+		Type:    "saptune-cmd-error",
+		Message: "error executing saptune command",
+	}
+)
+
+type SaptuneGatherer struct {
+	executor utils.CommandExecutor
+}
+
+func NewDefaultSaptuneGatherer() *SaptuneGatherer {
+	return NewSaptuneGatherer(utils.Executor{})
+}
+
+func NewSaptuneGatherer(executor utils.CommandExecutor) *SaptuneGatherer {
+	return &SaptuneGatherer{
+		executor: executor,
+	}
+}
+
+func (s *SaptuneGatherer) Gather(factsRequests []entities.FactRequest) ([]entities.Fact, error) {
+	cachedFacts := make(map[string]entities.Fact)
+
+	facts := []entities.Fact{}
+	log.Infof("Starting %s facts gathering process", SaptuneGathererName)
+	saptuneRetriever, err := saptune.NewSaptune(s.executor)
+
+	for _, factReq := range factsRequests {
+		var fact entities.Fact
+
+		internalArguments, ok := whitelistedArguments[factReq.Argument]
+		cachedFact, cacheHit := cachedFacts[factReq.Argument]
+
+		switch {
+		case err != nil:
```

Comment:
We don't need this `SaptuneNotInstalled` and `SaptuneVersionUnsupported` in the main loop.
If the `Gather` function return them as error, they will be added to all requests.

```
facts := []entities.Fact{}
log.Infof("Starting %s facts gathering process", SaptuneGathererName)
saptuneRetriever, err := saptune.NewSaptune(s.executor)
if err != nil {
    return facts, &SaptuneNotInstalled
}

if !saptuneRetriever.IsJSONSupported {
    return facts, &SaptuneVersionUnsupported
}

for _, factReq := range factsRequests {
...
```